glommio/src/read_write_lock.rs

```diff
@@ -0,0 +1,1214 @@
+/// on [`ReadWriteLock`].
+///
+/// [`write`]: ReadWriteLock::write
+/// [`try_write`]: ReadWriteLock::try_write
+#[must_use = "if unused the ReadWriteLock will immediately unlock"]
+#[derive(Debug)]
+pub struct ReadWriteLockWriteGuard<'a, T> {
+    rw: Rc<RefCell<State<T>>>,
+    //guard can not outlive rw-lock owner
+    phantom: PhantomData<&'a T>,
+}
+
+impl<'a, T> Deref for ReadWriteLockWriteGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        let state = (*self.rw).borrow();
+
+        if state.closed {
+            panic!("Related ReadWriteLock is already closed");
+        }
+
+        let state = (*self.rw).borrow();
+        let ptr = state.value.get();
+        unsafe { &*ptr }
+    }
+}
+
+impl<'a, T> DerefMut for ReadWriteLockWriteGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        let state = (*self.rw).borrow();
+
+        if state.closed {
+            panic!("Related ReadWriteLock is already closed");
+        }
+
+        let state = (*self.rw).borrow();
+        let ptr = state.value.get();
+        unsafe { &mut *ptr }
+    }
+}
+
+impl<'a, T> Drop for ReadWriteLockWriteGuard<'a, T> {
+    fn drop(&mut self) {
+        let mut state = self.rw.borrow_mut();
+
+        if !state.closed {
+            debug_assert!(state.writers > 0);
+            state.writers -= 1;
+
+            ReadWriteLock::<T>::wake_up_fibers(&mut state);
+        }
+    }
+}
+
+impl<T> ReadWriteLock<T> {
+    /// Creates a new instance of an `ReadWriteLock<T>` which is unlocked.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use glommio::read_write_lock::ReadWriteLock;
+    ///
+    /// let lock = ReadWriteLock::new(5);
+    /// ```
+    pub fn new(value: T) -> Self {
+        ReadWriteLock {
+            rw: Rc::new(RefCell::new(State::new(value))),
+        }
+    }
+
+    /// Returns a mutable reference to the underlying data.
+    ///
+    /// Since this call borrows the `ReadWriteLock` mutably, no actual locking needs to
+    /// take place -- the mutable borrow statically guarantees no locks exist.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the ReadWriteLock is closed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use glommio::read_write_lock::ReadWriteLock;
+    /// use glommio::LocalExecutor;
+    ///
+    /// let mut lock = ReadWriteLock::new(0);
+    /// let ex = LocalExecutor::make_default();
+    ///
+    /// ex.run(async move {
+    ///     *lock.get_mut().unwrap() = 10;
+    ///     assert_eq!(*lock.read().await.unwrap(), 10);
+    /// });
+    /// ```
+    pub fn get_mut(&mut self) -> ReadWriteLockResult<&mut T> {
+        let state = (*self.rw).borrow();
+        if state.closed {
+            return Err(LockClosedError);
+        }
+
+        Ok(unsafe { &mut *(state.value.get()) })
+    }
+
+    /// Locks this ReadWriteLock with shared read access, blocking the current fiber
```

Comment:
suggested saying "suspending" instead of blocking, as "blocking" has another meaning which is very loaded for us.

Reply:
Sure